portal-impl/src/com/liferay/portal/webserver/WebServerServlet.java

```diff
@@ -1374,6 +1374,19 @@ else if (pathArray.length == 2) {
 
 			// Unable to check with UUID because of multiple repositories
 
+			// Check if it is a folderId and not a UUID
+
+			long groupId = GetterUtil.getLong(pathArray[0], -1L);
+			long folderId = GetterUtil.getLong(pathArray[1], -1L);
+
+			if ((groupId > -1) && (folderId > -1)) {
+				if (_log.isDebugEnabled()) {
+					_log.debug(
+						"Trying to access a Folder, folderId=" + folderId);
+				}
+
+				throw new NoSuchFileEntryException();
+			}
```

Comment:
When `pathArray.length == 2` it means the URL should have the shape `/:groupId/:uuid`, not `/:folderId/:uuid`. Where are these URL with folder ID being generated? This shouldn't happen -- also, the reproduction steps in the bug clearly use a 3 part URL, so this fix doesn't make much sense. In addition to that, this is inconsistent with how the `_getFileEntry(String[])` method interprets these 2 part URLs.

@balazssk Can you post here the specific URLs that are triggering this problem? Can you also implement a solution that treats these URLs in the same way consistently?

Reply:
hi @adolfopa ,
The idea behind this is that users are subtracting `/:fileName` from `/:groupId/:folderId/:fileName/`, so the specific URL pattern is `/documents/:groupId/:folderId`. As far as I know, this should not be a URL generated by Liferay, nor should it be a valid one identifying a FileEntry.

> Can you also implement a solution that treats these URLs in the same way consistently?
Ideally, `_getFileEntry(String[])` handles this case, as the second part should be a UUID by design. If it is a folderId or any invalid String, `_getFileEntry` would simply trigger a `NoSuchFileEntryException` through `DLAppServiceUtil.getFileEntryByUuidAndGroupId`

Reply:
But what is the use case behind this? Users are randomly changing urls? Why?
In any case, if you need to validate URLs, this should be done in a more general way; this looks like a very specific fix for a very specific issue. Why not add a small layer of validation, and consistently check that URLs are valid?
Also, if _getFileEntry does the correct thing, why write totally different code to handle the same case? Why not use a similar approach?.

Reply:
Hi @adolfopa ,
Yes, users are randomly changing URLs in order to test if the custom 404 page is displayed or not.

> In any case, if you need to validate URLs, this should be done in a more general way; this looks like a very specific fix for a very specific issue.

I agree, but it looks like that document URLs only redirect to the wrong 404 page when 2 parts are present in the URL because VirtualHostFilter.isDocumentFriendlyURL uses this logic to determine if the FriendlyURLServlet or something else should handle it. This logic could be better if we could accertain that the 2nd param is a FileEntry uuid, but the comment found in the code indicates that is beyond our power.

> Why not add a small layer of validation, and consistently check that URLs are valid?

To be honest, I don't know where to begin designing something like that.

> Also, if `_getFileEntry` does the correct thing, why write totally different code to handle the same case? Why not use a similar approach?

That is a good question, I could not dig up any reasoning behind why `hasFiles` does not check 2 part URLs, maybe. Do you per chance have any insight on that?

Reply:
I don't have any insights of why it is that way, but can you go ahead and make the code for these two methods symmetrical? I'd ok with implementing a fix similar to what is done in _getFileEntry.

Reply:
cheers @adolfopa ,
I've opened https://github.com/liferay-lima/liferay-portal/pull/3766